consult-yasnippet.el

```diff
@@ -0,0 +1,140 @@
+;; the Free Software Foundation, either version 3 of the License, or
+;; (at your option) any later version.
+
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; Provides the command `consult-yasnippet' which presents all the snippet
+;; expansions available in the current buffer with in-buffer previews.
+;;
+;; This is an extra package, to avoid loading yasnippet in the base consult
+;; package..
+
+
+;;; Code:
+
+(require 'consult)
+(require 'yasnippet)
+
+(defun consult-yasnippet--expand-template (template region region-contents)
+  "Expand the yasnippet template TEMPLATE at point."
+  (deactivate-mark)
+  (goto-char (car region))
+  (when (not (string-equal "" region-contents))
+    (push-mark (point))
+    (push-mark (cdr region) nil t))
+
+  (yas-expand-snippet (yas--template-content template)
+                      nil nil
+                      (yas--template-expand-env template)))
+
+(defun consult-yasnippet--preview ()
+  "Previewer for `consult--read'.
+This function expands TEMPLATE at point in the buffer
+`consult-yasnippet--read-snippet' was started in. This includes
+overwriting any region that was active and removing any previous
+previews that're already active.
+
+When TEMPLATE is not given, this function essentially just resets
+the state of the current buffer to before any snippets were previewed."
+  (let* ((buf (current-buffer))
+         (region (if (region-active-p)
+                     (cons (region-beginning) (region-end))
+                   (cons (point) (point))))
+         (region-contents (buffer-substring (car region) (cdr region))))
+    (lambda (template restore)
+      (with-current-buffer buf
+        (let ((yas-verbosity 0)
+              (inhibit-redisplay t)
+              (inhibit-read-only t)
+              (orig-offset (- (point-max) (cdr region)))
+              (yas-prompt-functions '(yas-no-prompt)))
+
+          ;; We always undo any snippet previews before maybe setting up
+          ;; some new previews.
+          (delete-region (car region) (cdr region))
+          (goto-char (car region))
+          (setcar region (point))
+          (insert region-contents)
+          (setcdr region (point))
+
+          (when (and template (not restore))
+            (unwind-protect
+                (consult-yasnippet--expand-template template region region-contents)
+              (unwind-protect
+                  (mapc #'yas--commit-snippet
+                        (yas-active-snippets (point-min) (point-max)))
+                (setcdr region (- (point-max) orig-offset))))
+            (redisplay)))))))
```

Comment:
Why is this redisplay needed?

Reply:
No idea :yum:. It was there in ivy-yasnippet so I copied it verbatim.

Reply:
Yes sure, but this needs a bit of investigation. I will only pull code which is well-understood and explained. I must understand it at least, otherwise I cannot maintain it.

Reply:
That's understandable. I kind of assumed a lot of the stuff and logic in there is to deal with edge cases so I didn't think too hard about the ramifications. If you'd like to understand all the nitty gritty stuff, maybe you should ask @mkcms or @conao3, I'm sure they can be of more help.

Reply:
Yes, maybe @mkcms or @conao3 are interested in helping with a port of ivy-yasnippet to consult.

Reply:
I am the maintainer of ivy-yasnippet, but I am the maintainer of the current state and will not change the code unless there is a problem.
In that respect, I don't know the intention of this redisplay either. You can try removing it and see if it causes any problems in production.

(Actually, I became a fan of ivy-yasnippet as soon as I found the package, and I ran for maintainer in the absence of a maintainer, but there was a fatal glitch and I am not using it now.
 The bug is that if the snippet contains an Elisp that asks for input, when previewing the snippet in the buffer, it will override ivy-yasnippet and ask the user for input.
 If it didn't have this glitch it would be one reason for me to move from ivy to consult)

Reply:
@conao3 Thank you for chiming in here. Can you point us to a snippet which has this glitch? 

@mohkale It seems we have to deconstruct the code ourselves before it is ready for Consult. The presence of a known problematic glitch is reason enough to not merge it in the current form.

Reply:
@minad 

I don't really think there's a way around that on our end :/. Maybe we can ask the yasnippet maintainer to expose a function to expand a snippet, without any elisp being evaluated. Otherwise I guess we can leave this in limbo until we figure out a good way to handle it.

Thanks for all the time you've invested in this PR :+1:.

Reply:
Any snippet has `yas-completing-read`. Below snippet is a sample snippet I have.
```
# name: begin_src
# key: begin_src
# --
#+begin_src ${1: $$(yas-completing-read "Prompt: "
(append
 '("C++")
 (mapcar
  (lambda (elm)
    (substring (symbol-name elm) 3))
  (cl-remove-if-not
   (lambda (elm)
     (string-prefix-p "ob-" (symbol-name elm)))
   features))))}
  $0
#+end_src
```

For detail reprostep, please see https://github.com/mkcms/ivy-yasnippet/issues/10 .

Reply:
This does not sound like a deep problem to me. We can just cl-letf this yas read function locally returning some placeholder and go on.

Reply:
Is there only yas-completing-read or also yas-read-number etc? Which functions would require special handling? Alternatively we could just write our own preview expander assuming it is not too complicated.

Reply:
My concern is that if you can execute arbitrary lisp then nothing stops someone from running `completing-read` or any other blocking commands. Perhaps that's out of the concern for this particular problem. I'll try `cl-letf` and returning empty strings for everything.

Incidentally I can only see 3 yas commands with complete in them:
1. `yas-completing-read`
2. `yas-completing-prompt`
3. `yas/completing-prompt`

Reply:
Alright I've wrapped `yas-expand-snippet` in a `cl-letf` block such that `yas-completing-read` always returns an empty string. I'm now no longer getting any `completing-read` popups when previewing the snippet with `consult-yasnippet`.

Reply:
Okay, so that could be an option. But it wouldn't be 100% safe. It would be better to not execute elisp at all. How complicated is the expansion code? Maybe propose upstream to yasnippet a function which returns a snippet as previewable string. After it has been added there we can proceed here.

Reply:
I looked into yasnippet. Unfortunately it seems not to get much maintenance these days. For us, I see the following potential ways going forward, sorted by required effort:

1. Give up on the ivy-yasnippet/consult-yasnippet idea if it cannot be implemented reliably
2. Accept the unsafe solution (but this is not something I am eager to do honestly)
3. Write our own expander code here, which would work better in our scenario
5. Look for an alternative yasnippet-compatible/like package and support that instead (I think the alternatives are tempo, skeleton and yasnippet, maybe more?)
6. Try to contribute to upstream yasnippet, maybe they are even looking for a maintainer

Reply:
Is something like this good enough?

~~~ elisp
(defun my-expander ()
  (interactive)
  (message "%S"
           (mapcar (lambda (x)
                     (replace-regexp-in-string
                      "\\$[0-9]+" ""
                      (replace-regexp-in-string
                       "\\${[0-9]+:\\([^}]*\\)}" "\\1"
                       (yas--template-content (cdr x)))))                   
                   (car (yas--templates-for-key-at-point)))))
~~~

And then just show the obtained templates as an overlay as I originally suggested.

Reply:
@minad

I don't really have an issue with 2, I've been using ivy-yasnippet for years (exactly 2 :yum:) and I've never run into the issue that @conao3 mentioned. Although of course seing as the brunt of the maintanance work will be on you I understand your hesitation to do so.

If you'd like to go the 1 route that's fine. I can just add what I've got to the wiki or somewhere else and users will have to maintain it themselves. This is probably the best route for anything that doesn't directly conform to consults philosophy.

I find 3 interesting, but honestly not sure of the complexity or ramifications of it. If you can come up with a single function that takes a snippet template then returns a string then that's great :smile:, otherwise I'm not sure I'll be able to do so. In regards to using overlays there's still the issue with syntax highlighting rules when expanding snippets in memory or a temp-buffer. End of the day we need some way to apply the current buffers font-lock rules to it, the easiest of which would be to just insert it. (Maybe there's a function that can fontify a string, I'm not sure, I haven't found it yet).

As for 4, just a short while back yasnippet got a new patch allowing regexp keys (I believe) so it is still actively developed. I'm sure if you submitted a PR @joatavora would consider it :smile:.

Reply:
> I find 3 interesting, but honestly not sure of the complexity or ramifications of it. If you can come up with a single function that takes a snippet template then returns a string then that's great smile, otherwise I'm not sure I'll be able to do so. In regards to using overlays there's still the issue with syntax highlighting rules when expanding snippets in memory or a temp-buffer. End of the day we need some way to apply the current buffers font-lock rules to it, the easiest of which would be to just insert it. (Maybe there's a function that can fontify a string, I'm not sure, I haven't found it yet).

I agree. Overlays are a bad idea because of the fontification. I forgot this again. 

But I still think option 3 is the way to go. The way Consult is written as of now is such that previews are always "benign", they do nothing unexpected and must be solid under all circumstances. I think we should just use this simple string substitution function from above or something slightly more sophisticated. Maybe we can reuse some parser infrastructure from yasnippet? 

Would you like to continue to work on that or do you expect me to do this? I am willing to maintain this package as part of Consult. But I am not very interested in working out all the details. I already spent a bit of time on this PR, so it is obvious that I would like to have it here.

Reply:
The function I pasted above is a good starting point. But I think in order make this reliable we need something more sophisticated obviously.

~~~ elisp
(defun my-expander ()
  (interactive)
  (message "%S"
           (mapcar (lambda (x)
                     (replace-regexp-in-string
                      "\\$[0-9]+" ""
                      (replace-regexp-in-string
                       "\\${[0-9]+:\\([^}]*\\)}" "\\1"
                       (yas--template-content (cdr x)))))                   
                   (car (yas--templates-for-key-at-point)))))
~~~

Reply:
@minad

If you'd like to get started on this approach please do so. Otherwise I'm not sure when but I'll probably come back to this PR and try my hand at it when I feel the need for it. Atm the current implementation works for my needs so I'm not really in a rush to improve it IYGWIM. Regardless thank you for all the help you've given with this PR, I've learnt quite a lot :smile:.

Reply:
Okay, then I guess I close this PR for now. Maybe you or someone else will pick it up again in the future.